cpp/include/raft/mdarray.h

```diff
@@ -0,0 +1,408 @@
+
+  static_assert(!std::is_const<ElementType>::value,
+                "Element type for container must not be const.");
+
+  using pointer         = typename container_policy_type::pointer;
+  using const_pointer   = typename container_policy_type::const_pointer;
+  using reference       = typename container_policy_type::reference;
+  using const_reference = typename container_policy_type::const_reference;
+
+ private:
+  template <typename E,
+            typename ViewAccessorPolicy =
+              std::conditional_t<std::is_const_v<E>,
+                                 typename container_policy_type::const_accessor_policy,
+                                 typename container_policy_type::accessor_policy>>
+  using view_type_impl =
+    std::conditional_t<container_policy_type::is_host_type::value,
+                       host_mdspan<E, extents_type, layout_type, ViewAccessorPolicy>,
+                       device_mdspan<E, extents_type, layout_type, ViewAccessorPolicy>>;
+
+ public:
+  /**
+   * \brief the mdspan type returned by view method.
+   */
+  using view_type       = view_type_impl<element_type>;
+  using const_view_type = view_type_impl<element_type const>;
+
+ public:
+  constexpr mdarray() noexcept(std::is_nothrow_default_constructible<container_type>::value) =
+    default;
+  constexpr mdarray(mdarray const&) noexcept(
+    std::is_nothrow_copy_constructible<container_type>::value) = default;
+  constexpr mdarray(mdarray&&) noexcept(std::is_nothrow_move_constructible<container_type>::value) =
+    default;
+
+  auto operator =(mdarray&&) noexcept(std::is_nothrow_move_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  auto operator =(mdarray const&) noexcept(std::is_nothrow_copy_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  ~mdarray() noexcept(std::is_nothrow_destructible<container_type>::value) = default;
+
+  /**
+   * @brief The only constructor that can create storage, this is to make sure CUDA stream is being
+   * used.
+   */
+  constexpr mdarray(mapping_type const& m, container_policy_type const& cp)
+    : cp_(cp), map_(m), c_(cp_.create(map_.required_span_size()))
+  {
+  }
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  auto view() noexcept { return view_type(c_.data(), map_, cp_.view()); }
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  auto view() const noexcept { return const_view_type(c_.data(), map_, cp_.view()); }
+
+  [[nodiscard]] constexpr auto size() const noexcept -> size_type { return this->view().size(); }
+
+  auto data() noexcept -> pointer { return c_.data(); }
+  constexpr auto data() const noexcept -> const_pointer { return c_.data(); }
+
+  /**
+   * @brief Indexing operator, use it sparingly since it triggers a device<->host copy.
+   */
+  template <typename... IndexType>
+  auto operator()(IndexType&&... indices)
```

Comment:
Thank you for the suggestion,  Currently, the mdarray class leaves the cuda stream to container policy.   Adding an extra stream parameter might cause conflict with the stream in the policy.  I'm not entirely sure whether do we need an extra stream for accessing the operator.

Reply:
Then, maybe we should add `view()` overloads that would accept a new container policy as an argument? So that it would be easy to construct different views per-stream.